File: Javascript/Design/M-1268. Search Suggestions System.ts
class ProductTrieNode {
  children: object;
  products: string[];
  constructor() {
    this.children = {};
    this.products = [];
  }
}

class ProductTrie {
  root: ProductTrieNode;

  constructor(products: string[]) {
    this.root = new ProductTrieNode();
    this.initProductTrie(products.sort());
  }
  initProductTrie(products: string[]): void {
    for (const product of products) {
      let node = this.root;
      for (const char of product) {
        if (!(char in node.children)) {
          node.children[char] = new ProductTrieNode();
        }
        node = node.children[char];
        if (node.products.length < 3) node.products.push(product);
      }
    }
  }
  getProductsBySearchWord(searchWord: string): string[] {
    let node = this.root;
    for (const char of searchWord) {
      if (!(char in node.children)) return [];
      node = node.children[char];
    }
    return node.products;
  }
}

function suggestedProducts(products: string[], searchWord: string): string[][] {
  const productTrie = new ProductTrie(products);
  const result: string[][] = [];
  let typedText = '';
  for (const char of searchWord) {
    typedText += char;
    result.push(productTrie.getProductsBySearchWord(typedText));
  }

  return result;
}

const testData = [
  {
    products: ['mobile', 'mouse', 'moneypot', 'monitor', 'mousepad'],
    searchWord: 'mouse',
    ans: [
      ['mobile', 'moneypot', 'monitor'],
      ['mobile', 'moneypot', 'monitor'],
      ['mouse', 'mousepad'],
      ['mouse', 'mousepad'],
      ['mouse', 'mousepad'],
    ],
  },
  {
    products: ['havana'],
    searchWord: 'havana',
    ans: [
      ['havana'],
      ['havana'],
      ['havana'],
      ['havana'],
      ['havana'],
      ['havana'],
    ],
  },
];

for (const data of testData) {
  console.log(
    JSON.stringify(data.ans) ===
      JSON.stringify(suggestedProducts(data.products, data.searchWord))
  );
}

File: Design/M-1268. Search Suggestions System/M-1268. Search Suggestions System.py
class ProductTrieNode:
    def __init__(self) -> None:
        self.children = {}
        self.products: list[str] = []


class ProductTrie:
    def __init__(self, products: list[str]) -> None:
        self.root = ProductTrieNode()
        products.sort()
        self.init_product_trie(products)

    def init_product_trie(self, products: list[str]) -> None:
        for product in products:
            node = self.root
            for char in product:
                if char not in node.children:
                    node.children[char] = ProductTrieNode()
                node: ProductTrieNode = node.children[char]
                if len(node.products) < 3:
                    node.products.append(product)

    def get_products_by_word(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                return []
            node = node.children[char]
        return node.products


class Solution:
    @classmethod
    def suggestedProducts(
        self, products: list[str], searchWord: str
    ) -> list[list[str]]:
        product_trie = ProductTrie(products)
        result: list[list[str]] = []
        typed_text = ""
        for char in searchWord:
            typed_text += char
            result.append(product_trie.get_products_by_word(typed_text))
        return result


testData = [
    {
        "products": ["mobile", "mouse", "moneypot", "monitor", "mousepad"],
        "searchWord": "mouse",
        "ans": [
            ["mobile", "moneypot", "monitor"],
            ["mobile", "moneypot", "monitor"],
            ["mouse", "mousepad"],
            ["mouse", "mousepad"],
            ["mouse", "mousepad"],
        ],
    },
    {
        "products": ["havana"],
        "searchWord": "havana",
        "ans": [
            ["havana"],
            ["havana"],
            ["havana"],
            ["havana"],
            ["havana"],
            ["havana"],
        ],
    },
]

for data in testData:
    print(
        data["ans"] == Solution.suggestedProducts(data["products"], data["searchWord"])
    )
